query/query.go

```diff
@@ -2395,6 +2414,34 @@ func ProcessGraph(ctx context.Context, sg, parent *SubGraph, rch chan error) {
 	rch <- childErr
 }
 
+// applies "random" to lists inside uidMatrix
+// Params.Random number of nodes are selected
+// duplicates are not removed i.e., random selection by replacement is done
```

Comment:
We should not have duplicate results. I'd suggest having some map to keep track of what all we have already picked to randomly pick `n` unique results.

Reply:
Okay, I'll do this, along with the filters application.

Reply:
I don't think you need to do anything for filters. You are already doing the random selection after filter application and pagination.

Reply:
Oh okay. That explains why I couldn't figure out how to do it by looking at pagination xD.

Reply:
I have pushed a commit fixing these. 

To remove duplicates, I had to:
 - introduce a new struct `UidKey` to store the index of a UID in the matrix
 - calculate total number of nodes and check if required number of nodes is less than that. If it's not, return all the nodes.

Because of the condition `if sg.Params.Random >= totalNodes` at line 2440, when the required number of nodes (`sg.Params.Random`) is equal to the total number of nodes, all of the nodes are returned without any processing. I did this thinking of it as an optimization (because there will be many overlaps in random numbers when only a few nodes are left to be selected). But I just realized that this will make it so that the **order is not randomized when `sg.Params.Random == totalNodes`**. Should I fix this to do the random selection even when all the nodes have to be selected?